rust/routee-compass-core/src/model/state/state_model.rs

```diff
@@ -0,0 +1,358 @@
+use super::{
+    state_error::StateError, state_feature::StateFeature, state_model_entry::StateModelEntry,
+    update_operation::UpdateOperation,
+};
+use crate::model::traversal::state::state_variable::StateVar;
+use itertools::Itertools;
+use serde_json::json;
+use std::collections::HashMap;
+
+pub struct StateModel(HashMap<String, StateModelEntry>);
```

Comment:
This can be a future work idea but I tested out storing 3 state model entries as a vector and then just scanning the vector to find the right entry and it's about 4x faster to get the last entry and 8x faster to get the first entry (over 1M iterations). I wonder if we could either default to storing the entries as a vector that we scan or expand the internal store such that if we have a large number of state variables (maybe >10) than we use a hash map but if we have a small number than we use a vector.

Reply:
totally. that's how [Scala implements `Map`. for Map instances with 4 or less elements](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/Map.scala#L259), they rely on a scan, and don't actually instantiate a hashmap until 5+. that could make sense, and perhaps we pull it off by turning StateModel into:

```rust
pub enum StateModel {
  OneFeature {
    name: String,
    entry: StateModelEntry
  },
  TwoFeatures { 
    name1: String,
    name2: String,
    entry1: StateModelEntry,
    entry2: StateModelEntry,
  },
  ...
  NFeatures(HashMap<String, StateModelEntry>),
}
```

Reply:
yeah that's a good idea, and then getting it would be simple and performant like:

```rust
fn get(&self, name: &str) -> StateModelEntry {
 match self {
   StateModel::OneFeature => self.entry,
   StateModel::TwoFeatures => {
    if self.name1 == name {
     self.entry1
    } else {
     self.entry2
    },
    // and so forth
   }
 }
}
```